frame/contracts/src/tests.rs

```diff
@@ -1867,10 +2157,653 @@ fn ecdsa_recover() {
+			<Error<Test>>::StorageDepositNotEnoughFunds,
+		);
+
+		assert_eq!(System::events(), vec![]);
+	});
+}
+
+#[test]
+fn remove_code_works() {
+	let (wasm, code_hash) = compile_module::<Test>("dummy").unwrap();
+
+	ExtBuilder::default().existential_deposit(100).build().execute_with(|| {
+		let _ = Balances::deposit_creating(&ALICE, 1_000_000);
+
+		// Drop previous events
+		initialize_block(2);
+
+		assert_ok!(Contracts::upload_code(
+			Origin::signed(ALICE),
+			wasm,
+			Some(codec::Compact(1_000))
+		));
+
+		assert!(<CodeStorage<Test>>::contains_key(code_hash));
+		assert_ok!(Contracts::remove_code(Origin::signed(ALICE), code_hash));
+		assert!(!<CodeStorage<Test>>::contains_key(code_hash));
+
+		assert_eq!(
+			System::events(),
+			vec![
+				EventRecord {
+					phase: Phase::Initialization,
+					event: Event::Balances(pallet_balances::Event::Reserved {
+						who: ALICE,
+						amount: 180,
+					}),
+					topics: vec![],
+				},
+				EventRecord {
+					phase: Phase::Initialization,
+					event: Event::Contracts(crate::Event::CodeStored { code_hash }),
+					topics: vec![],
+				},
+				EventRecord {
+					phase: Phase::Initialization,
+					event: Event::Balances(pallet_balances::Event::Unreserved {
+						who: ALICE,
+						amount: 180,
+					}),
+					topics: vec![],
+				},
+				EventRecord {
+					phase: Phase::Initialization,
+					event: Event::Contracts(crate::Event::CodeRemoved { code_hash }),
+					topics: vec![],
+				},
+			]
+		);
+	});
+}
+
+#[test]
+fn remove_code_wrong_origin() {
+	let (wasm, code_hash) = compile_module::<Test>("dummy").unwrap();
+
+	ExtBuilder::default().existential_deposit(100).build().execute_with(|| {
+		let _ = Balances::deposit_creating(&ALICE, 1_000_000);
+
+		// Drop previous events
+		initialize_block(2);
+
+		assert_ok!(Contracts::upload_code(
+			Origin::signed(ALICE),
+			wasm,
+			Some(codec::Compact(1_000))
+		));
+
+		assert_noop!(
+			Contracts::remove_code(Origin::signed(BOB), code_hash),
+			sp_runtime::traits::BadOrigin,
+		);
+
+		assert_eq!(
+			System::events(),
+			vec![
+				EventRecord {
+					phase: Phase::Initialization,
+					event: Event::Balances(pallet_balances::Event::Reserved {
+						who: ALICE,
+						amount: 180,
+					}),
+					topics: vec![],
+				},
+				EventRecord {
+					phase: Phase::Initialization,
+					event: Event::Contracts(crate::Event::CodeStored { code_hash }),
+					topics: vec![],
+				},
+			]
+		);
+	});
+}
+
+#[test]
+fn remove_code_in_use() {
+	let (wasm, code_hash) = compile_module::<Test>("dummy").unwrap();
+
+	ExtBuilder::default().existential_deposit(100).build().execute_with(|| {
+		let _ = Balances::deposit_creating(&ALICE, 1_000_000);
+
+		assert_ok!(Contracts::instantiate_with_code(
+			Origin::signed(ALICE),
+			0,
+			GAS_LIMIT,
+			None,
+			wasm,
+			vec![],
+			vec![],
+		));
+
+		// Drop previous events
+		initialize_block(2);
+
+		assert_noop!(
+			Contracts::remove_code(Origin::signed(BOB), code_hash),
```

Comment:
```suggestion
			Contracts::remove_code(Origin::signed(ALICE), code_hash),
```

As per previous test, only origin can remove the code.